File: class22/java/TCC2201-TabelaSimbolos.java
// TCC2201-TabelaSimbolos.java | Aula 22
// Tabela de símbolos reutilizável para o compilador.
// Espelho Java do TCC2201-TabelaSimbolos.py (mesmas operações e saídas).
import java.util.*;

public class TCC2201TabelaSimbolos {

    static class Simbolo {
        String nome, tipo;
        Object valor;
        Simbolo(String nome, String tipo, Object valor) {
            this.nome = nome; this.tipo = tipo; this.valor = valor;
        }
        public String toString() { return nome + ":" + tipo + "=" + valor; }
    }

    static class TabelaSimbolos {
        Map<String, Simbolo> simbolos = new LinkedHashMap<>();

        Simbolo declarar(String nome, String tipo, Object valor) {
            if (simbolos.containsKey(nome))
                throw new RuntimeException("Variável '" + nome + "' já declarada");
            Simbolo sim = new Simbolo(nome, tipo, valor);
            simbolos.put(nome, sim);
            return sim;
        }

        Simbolo buscar(String nome) { return simbolos.get(nome); }

        Simbolo atribuir(String nome, Object valor) {
            Simbolo sim = simbolos.get(nome);
            if (sim == null)
                throw new RuntimeException("Variável '" + nome + "' não declarada");
            sim.valor = valor;
            return sim;
        }

        public String toString() {
            StringBuilder sb = new StringBuilder("TabelaSimbolos(");
            boolean first = true;
            for (Simbolo s : simbolos.values()) {
                if (!first) sb.append(", ");
                sb.append(s);
                first = false;
            }
            return sb.append(")").toString();
        }
    }

    public static void main(String[] args) {
        TabelaSimbolos ts = new TabelaSimbolos();
        ts.declarar("x", "int", 42);
        ts.declarar("y", "float", 3.14);
        ts.declarar("nome", "string", "Giovane");

        System.out.println("=== Tabela após declarações ===");
        System.out.println(ts);

        System.out.println("\n=== Atribuições ===");
        ts.atribuir("x", 100);
        ts.atribuir("y", 2.71);
        System.out.println(ts);

        System.out.println("\n=== Buscas ===");
        System.out.println("  x = " + ts.buscar("x"));
        System.out.println("  y = " + ts.buscar("y"));

        System.out.println("\n=== Erros ===");
        System.out.println("  z = " + ts.buscar("z") + " (null — não encontrado)");
        try {
            ts.declarar("x", "int", 0);
        } catch (RuntimeException e) {
            System.out.println("  " + e.getMessage());
        }
    }
}

File: class22/python/TCC2201-TabelaSimbolos.py
#!/usr/bin/env python3
"""TCC2201-TabelaSimbolos.py | Aula 22
Tabela de símbolos reutilizável para o compilador.
Armazena: nome, tipo, valor.
"""
from dataclasses import dataclass, field
from typing import Any, Dict, Optional


@dataclass
class Simbolo:
    nome: str
    tipo: str = "int"
    valor: Any = None

    def __repr__(self):
        return f"{self.nome}:{self.tipo}={self.valor}"


class TabelaSimbolos:
    def __init__(self):
        self._simbolos: Dict[str, Simbolo] = {}

    def declarar(self, nome: str, tipo: str = "int", valor: Any = None) -> Simbolo:
        if nome in self._simbolos:
            raise NameError(f"Variável '{nome}' já declarada")
        sim = Simbolo(nome, tipo, valor)
        self._simbolos[nome] = sim
        return sim

    def buscar(self, nome: str) -> Optional[Simbolo]:
        return self._simbolos.get(nome)

    def atribuir(self, nome: str, valor: Any) -> Simbolo:
        sim = self._simbolos.get(nome)
        if sim is None:
            raise NameError(f"Variável '{nome}' não declarada")
        sim.valor = valor
        return sim

    def __contains__(self, nome: str) -> bool:
        return nome in self._simbolos

    def __repr__(self):
        return "TabelaSimbolos(" + ", ".join(str(s) for s in self._simbolos.values()) + ")"


if __name__ == "__main__":
    ts = TabelaSimbolos()
    ts.declarar("x", "int", 42)
    ts.declarar("y", "float", 3.14)
    ts.declarar("nome", "string", "Giovane")

    print("=== Tabela após declarações ===")
    print(ts)

    print("\n=== Atribuições ===")
    ts.atribuir("x", 100)
    ts.atribuir("y", 2.71)
    print(ts)

    print("\n=== Buscas ===")
    print(f"  x = {ts.buscar('x')}")
    print(f"  y = {ts.buscar('y')}")

    print("\n=== Erros ===")
    try:
        ts.buscar("z").valor
    except AttributeError:
        print("  z não encontrado (None)")
    try:
        ts.declarar("x")
    except NameError as e:
        print(f"  {e}")
